bisection.go

```diff
@@ -57,7 +57,20 @@ func (b *Bisection) Init(loc LinesearchLocation, step float64, _ *FunctionInfo)
 }
 
 func (b *Bisection) Finished(loc LinesearchLocation) bool {
-	return StrongWolfeConditionsMet(loc.F, loc.Derivative, b.initF, b.initGrad, b.currStep, 0, b.GradConst)
+	// Don't finish the linesearch until a minimum is found that is better than
+	// the best point found so far. We want to end up in the lowest basin of
+	// attraction
+	minF := b.initF
+	if b.maxF < b.initF {
+		minF = b.maxF
+	}
+	if b.minF < minF {
+		minF = b.minF
+	}
+	if StrongWolfeConditionsMet(loc.F, loc.Derivative, minF, b.initGrad, b.currStep, 0, b.GradConst) {
```

Comment:
I may not see through your intention here, but replacing initF with the minimum of initF, minF and maxF but keeping initGrad seems to me like a possible cause of trouble. For Armijo condition, you are essentially asking for initGrad-proportional decrease from minF value, which might be impossible to achieve, especially close to a minimum and/or for difficult functions. This problem might be more prominent with CG, because the directions are not well-scaled and not so good (just guessing here, I haven't go around to test it). Am I missing something?